signer/tests/integration/transaction_coordinator.rs

```diff
@@ -3413,3 +3413,239 @@ async fn test_conservative_initial_sbtc_limits() {
+
+        // Write the blocks to the database.
+        db.write_blocks(&bitcoin_chain, &stacks_chain).await;
+
+        // Get the chain tips and assert that they're what we expect.
+        let (bitcoin_chain_tip, stacks_chain_tip) = db.get_chain_tips().await;
+        assert_eq!(
+            bitcoin_chain_tip.block_hash,
+            bitcoin_chain.chain_tip().block_hash
+        );
+        assert_eq!(stacks_chain_tip, stacks_chain.chain_tip().block_hash);
+
+        // Create DKG shares and write them to the database.
+        let dkg_shares = model::EncryptedDkgShares {
+            aggregate_key: signer_set.aggregate_key(),
+            started_at_bitcoin_block_hash: bitcoin_chain_tip.block_hash,
+            started_at_bitcoin_block_height: bitcoin_chain_tip.block_height,
+            signer_set_public_keys: signer_set.signer_keys().to_vec(),
+            dkg_shares_status: DkgSharesStatus::Verified,
+            ..Faker.fake()
+        };
+        db.write_encrypted_dkg_shares(&dkg_shares).await.unwrap();
+
+        (bitcoin_chain, stacks_chain, signer_set, signer_keys)
+    }
+
+    struct TestParams {
+        chain_length: u64,
+        signature_threshold: u16,
+        sbtc_limits: SbtcLimits,
+        amount: u64,
+        num_approves: usize,
+        num_expected_results: usize,
+        at_block_height: usize,
+    }
+
+    /// Asserts that TxCoordinatorEventLoop::get_pending_requests processes
+    /// withdrawals
+    #[test_case(TestParams {
+        signature_threshold: 2,
+        num_approves: 3,
+        amount: 1_000,
+        sbtc_limits: SbtcLimits::unlimited(),
+        chain_length: 8,
+        at_block_height: 2,
+        num_expected_results: 1,
+    }; "should_pass_all_validations")]
+    #[test_case(TestParams {
+        signature_threshold: 2,
+        num_approves: 3,
+        amount: 100,
+        sbtc_limits: SbtcLimits::unlimited(),
+        chain_length: 8,
+        at_block_height: 2,
+        num_expected_results: 0,
+    }; "below_dust_limit")]
+    #[test_case(TestParams {
+        signature_threshold: 2,
+        num_approves: 3,
+        amount: 1_000,
+        sbtc_limits: SbtcLimits::zero(),
+        chain_length: 8,
+        at_block_height: 2,
+        num_expected_results: 0,
+    }; "amount_over_per_withdrawal_limit")]
+    #[test_case(TestParams {
+        signature_threshold: 2,
+        num_approves: 3,
+        amount: 1_000,
+        sbtc_limits: SbtcLimits::unlimited(),
+        chain_length: 5,
+        at_block_height: 0,
+        num_expected_results: 0,
+    }; "insufficient_confirmations")]
+    #[test_case(TestParams {
+        signature_threshold: 2,
+        num_approves: 1,
+        amount: 1_000,
+        sbtc_limits: SbtcLimits::unlimited(),
+        chain_length: 8,
+        at_block_height: 2,
+        num_expected_results: 0,
+    }; "insufficient_votes")]
+    #[test_log::test(tokio::test)]
+    async fn test_validations(params: TestParams) {
```

Comment:
Nice, it's not easy getting to the place where the tests are easy two follow but require so much setup. Good job.

Reply:
Yeah I'm still not in love with it; I would prefer to test each validation rule separately and have a better assertion, but time got to me so I manually verified the log output for each case instead for now (to make sure it was the correct skip reason).